### ISSUE TITLE
Resolve some naming questions

### ISSUE THREAD

**Issue by dsyme** [xamarin/TorchSharp#225]
1. I'm not particularly tied to `open TorchSharp.Tensor` namespace.  I'd be happy with `open TorchSharp` or anything that gives the ergonomics you want

2. When updating DiffSharp to latest I noticed `TorchSharp.Device` rather than `TorchSharp.TorchDevice`.  Previously I was putting `Torch` on the front to help disambiguate w.r.t. DiffSharp "Tensor".  But I understand you'll likely need to switch to plain `Tensor`.  But either do one or the other

        TorchSharp.DeviceType
        TorchSharp.Device
        TorchSharp.Tensor
        TorchSharp.Scalar

or

        TorchSharp.TorchDeviceType
        TorchSharp.TorchDevice
        TorchSharp.TorchTensor
        TorchSharp.TorchScalar

Right now we're a mix of these.

**Comment by NiklasGustafsson** [comment]
My preference would be to remove the 'Torch' prefix -- it is essentially a redundant encoding of the namespace in type names.

**Comment by NiklasGustafsson** [comment]
How about:

Torch.DeviceType
Torch.Device
Torch.Tensor
Torch.Scalar

**Comment by dsyme** [comment]
This is ok by me.

**Comment by NiklasGustafsson** [comment]
So, we change 'TorchSharp' to 'Torch' and remove the prefix on all types where it differs from PyTorch? 

Do we capitalize following traditional .NET conventions, or do we try to stay as close to PyTorch as possible? Tensorflow.NET has made a consistent effort to stay close to PyTorch, which I believe is a SciSharp design principle. Do we do the same here?

**Comment by dsyme** [comment]
> So, we change 'TorchSharp' to 'Torch' and remove the prefix on all types where it differs from PyTorch?

Actually TBH I'd probably prefer "TorchSharp" since that's the name of the tech.  

> Do we capitalize following traditional .NET conventions, or do we try to stay as close to PyTorch as possible? Tensorflow.NET has made a consistent effort to stay close to PyTorch, which I believe is a SciSharp design principle. Do we do the same here?

We stay as close as possible to PyTorch.  But for type names we use capitalization.  See https://github.com/xamarin/TorchSharp/issues/226 for the broader question of how to get closer to PyTorch through a lower-case `torch` type.    Indeed for this reason I think it's better to keep `open TorchSharp` as the very top namespace so that approach wouldn't lead to confusion

**Comment by NiklasGustafsson** [comment]
Method/function names are a jolly blend of Pascal- and all-lowercase names. Should we systematically go and change all that to be as similar to PyTorch as possible?

**Comment by dsyme** [comment]
> TorchSharp method/function names that correspond to PyTorch entries are a blend of Pascal- and all-lowercase names. Should we systematically go and change all that to be as similar to PyTorch as possible?

Yes :) 

> PyTorch namespaces are lower-case, so it should be 'torch.nn', not 'TorchSharp.NN' -- should we go that far?

No, not the namespaces.  I think SciSharp doesn't go that far, and it's right not to.

**Comment by GeorgeS2019** [comment]
@dsyme @NiklasGustafsson 
For me the eventual aim is to have a .NET to HuggingFace which is based on TorchSharp. HuggingFace has both tensorflow and pytorch implementation. It seems, for rust and go, torch API way is more popular. 

Currently, there is no satisfying ONNX C# API to address the lack of .NET approach to the latest transformer Deep AI innovation. This will put .NET developers a server handicap. 

Thanks for pushing the completion of TorchSharp API.

I hope the goal is to address the severe gap for .NET developers accessing the latest innovaiton in transformer Deep AI illustrated by HuggingFace through pytorch implementation.

==> it is hope that the naming decision will be one that will speed up the porting of python to .NET Deep AI implemented through Torch.

**Comment by lostmsu** [comment]
Re: @GeorgeS2019

This is out of scope of the particular issue, and of TorchSharp, but my commercial binding to TensorFlow [handled GPT-2 since about 1 month after OpenAI released it](https://habr.com/en/post/453232/). I even recently published GPT-2 as a library: https://github.com/losttech/GPT. The full TensorFlow API is exposed, so porting SOTA open source models takes no time. Did [SIREN](https://github.com/losttech/Siren) in a couple of days.

Frankly, .NET team seems to lack interest in taking on Python's reign in deep learning. I pitched them this project 2 years ago and they were dead set on just exposing low-level C API. .NET interactive folks are [deaf to our threading issues with TensorFlow](https://github.com/dotnet/interactive/issues/24). At least 64-bit arrays are starting to get some traction (maybe, post 6.0).

Sorry for the rant.

**Comment by GeorgeS2019** [comment]
@lostmsu First, I follow your work on GPT-2 using python Tensorflow wrapped through PythonNET (hence full Tensorflow API), an approach that many **NOW** see that as an intermediate step to industry standard for deployment. I congratulate U as **the first one having the vision to deliver GPT-2** to .NET community. 

Perhaps U were 2 years ahead of your time. Perhaps NOW is the time. 

The .NET community needs a C API way urgently to torch and tensorflow. Without which, there is a risk that C# as a language will be made obsolete in the race towards  enterprise wide AI need. 

**Do not give up, but be open minded** that there are many paths. **Each one of us** is doing our part so the main .NET programming and tool set we have been using over a few decades do not becoming obsolete in the process of keeping up with the fast pace of latest AI development.

**Comment by dsyme** [comment]
@lostmsu With regard to https://github.com/dotnet/interactive/issues/24, we should just determine a fix together, propose it and iterate to get it accepted into .NET Interactive.  It is open source and we should determine the change we want.  The code isn't that hard to modify tbh.

**Comment by NiklasGustafsson** [comment]
I had a conversation with @dyme offline, and here's a proposal:

We will strive to make the TorchSharp experience as close to Python as possible. Among other things, this means not always following the standard .NET naming conventions. 

Specifically, namespaces and classes corresponding to Python modules, will have lower-case names. Function and parameter names may have C-style casing, using underscores as word separators, instead of camel or Pascal casing.

Since namespaces cannot directly contain methods, we will continue to rely on extensive use of static classes to parallel Python modules.

We will deprecate, and eventually eliminate, the element-type-specific tensor factories, i.e. `Int32Tensor`, `Float32Tensor`, `BoolTensor`, etc. Instead, we will use the same mechanisms as Python does, but take advantage of static typing as appropriate.

For example, to create a 32-bit float tensor with random numbers, you will use `torch.rand(...)`. By either passing in the element type, or relying on the default `dtype` set in the runtime.

Properties will, as necessary, be renamed to correspond to the PyTorch names. For example, `ndims` instead of `dims` or `Dimensions`

Increased reliance on overloading. Pytorch has a lot of APIs where an argument can be an integer, float, a tuple or list, or a string. TorchSharp already relies on overloading in some cases, but there are more opportunities to do so, and we will attempt to systematically do the work to give users the same flexibility as in Python, while still taking full advantage of .NET's type system.

**Comment by NiklasGustafsson** [comment]
**Note; the proposed changes will be a massive breaking change.**
